### PR TITLE
Fix CMake golang rebuild problems

### PR DESCRIPTION
**Pull request by reyoung**
* In previous cmake, golang target will always rebuild no matter
go sources are changed or not.

### CODE REVIEW

File: cmake/generic.cmake [hunk 1]
```diff
@@ -306,11 +306,14 @@ function(go_library TARGET_NAME)
...
   set(dummyfile ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_NAME}_dummy.c)
 
+  file(GLOB GO_SOURCE RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}" "*.go")
+  set(TOUCH_CMD ${PROJ_ROOT}/paddle/scripts/build/if_newer_then_touch.sh)
   # This custom command will always run since it depends on a not
   # existing file.
   add_custom_command(
     OUTPUT dummy_rebulid_${TARGET_NAME}
-    COMMAND cmake -E touch ${dummyfile}
+    COMMAND ${TOUCH_CMD} ${dummyfile} ${GO_SOURCE}
```
Comment (wangkuiyi) [inline cmake/generic.cmake]: If we change this line into 

```
COMMAND ${PROJ_ROOT}/paddle/scripts/build/touch_if_newer.sh ${dummyfile} ${GO_SOURCE}
```

we can remove L310 which defines `TOUCH_CMD`.

File: paddle/scripts/build/if_newer_then_touch.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (wangkuiyi) [inline paddle/scripts/build/if_newer_then_touch.sh]: I think we need a comment here

```
# Usage: 
#    touch_if_newer.sh <reference_file> <source_file_1> <source_file_2> ...
# Touch a source file if it is newer than the reference file.
```

**Review by wangkuiyi — APPROVED** [review]
It would be helpful if we can have an issue that describes the problem encountered and the reason, and mark this PR as fixing the issue.

**Review by helinwang — CHANGES_REQUESTED** [review]
Thanks for trying to fix this!

There is no easy solution to this problem, since the Go source files is not just the ones in the specified folder. All the dependencies in other folders are the source files are well.

Rebuilding every time is done intentionally since there is no easy fix, and we can't risk not building the code when the source have changed.

However, even if we rebuild every time, Go already caches the build result (for all the source that is in $GOPATH), so rebuild should be very fast.

**Pull request closed without merging**